coderd/workspaceapps/stats.go

```diff
@@ -0,0 +1,378 @@
+			r := *stat
+			if r.SessionEndedAt.IsZero() {
+				// Report an end time for incomplete sessions, it will
+				// be updated later. This ensures that data in the DB
+				// will have an end time even if the service is stopped.
+				r.SessionEndedAt = sc.opts.Now().UTC() // Use UTC like database.Now().
+			}
+			report = append(report, r) // Report it (ended or incomplete).
+			if stat.SessionEndedAt.IsZero() {
+				newGroup = append(newGroup, stat) // Keep it for future updates.
+			}
+		}
+		if rollupChanged {
+			report = append(report, *rolledUp)
+		}
+
+		// Future rollups should only consider the compacted group.
+		sc.groupedStats[g] = newGroup
+
+		// Keep the group around until the next rollup window has passed
+		// in case data was collected late.
+		if len(newGroup) == 1 && rolledUp.SessionEndedAt.Add(sc.opts.RollupWindow).Before(sc.opts.Now()) {
+			delete(sc.groupedStats, g)
+		}
+	}
+
+	return report
+}
+
+func (sc *StatsCollector) flush(ctx context.Context) (err error) {
+	sc.opts.Logger.Debug(ctx, "flushing workspace app stats")
+	defer func() {
+		if err != nil {
+			sc.opts.Logger.Error(ctx, "failed to flush workspace app stats", "error", err)
+		} else {
+			sc.opts.Logger.Debug(ctx, "flushed workspace app stats")
+		}
+	}()
+
+	// We keep the backlog as a simple slice so that we don't need to
+	// attempt to merge it with the stats we're about to report. This
+	// is because the rollup is a one-way operation and the backlog may
+	// contain stats that are still in the statsBySessionID map and will
+	// be reported again in the future. It is possible to merge the
+	// backlog and the stats we're about to report, but it's not worth
+	// the complexity.
+	if len(sc.backlog) > 0 {
+		err = sc.opts.Reporter.Report(ctx, sc.backlog)
+		if err != nil {
+			return xerrors.Errorf("report workspace app stats from backlog failed: %w", err)
+		}
+		sc.backlog = nil
+	}
+
+	stats := sc.rollup()
+	if len(stats) == 0 {
+		return nil
+	}
+
+	err = sc.opts.Reporter.Report(ctx, stats)
+	if err != nil {
+		sc.backlog = stats
+		return xerrors.Errorf("report workspace app stats failed: %w", err)
+	}
+
+	return nil
+}
+
+func (sc *StatsCollector) Close() error {
+	sc.cancel()
+	<-sc.done
+	return nil
+}
+
+func (sc *StatsCollector) start() {
+	defer func() {
+		close(sc.done)
+		sc.opts.Logger.Info(sc.ctx, "workspace app stats collector stopped")
+	}()
+	sc.opts.Logger.Info(sc.ctx, "workspace app stats collector started")
+
+	ticker := time.NewTicker(sc.opts.ReportInterval)
+	defer ticker.Stop()
+
+	var reportFlushDone chan<- struct{}
+	done := false
+	for !done {
+		select {
+		case <-sc.ctx.Done():
+			ticker.Stop()
+			done = true
+		case <-ticker.C:
+		case reportFlushDone = <-sc.opts.Flush:
+		}
+
+		// Ensure we don't hold up this request for too long. Add a few
+		// seconds to prevent very short intervals from causing a timeout.
+		ctx, cancel := context.WithTimeout(context.Background(), sc.opts.ReportInterval+5*time.Second)
```

Comment:
I would just use a fixed timeout like 15s or something, and then call `ticker.Reset` after so the interval is the duration between the end of one flush and the beginning of another flush.